File: tf2-item-scanner/src/com/minowak/scanner/engine/Backpack.java
package com.minowak.scanner.engine;

import java.util.LinkedList;
import java.util.List;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import com.minowak.scanner.schema.ItemQuality;
import com.minowak.scanner.schema.TF2Item;
import com.minowak.scanner.utils.Configuration;

public class Backpack extends SteamEntity {
	private String apiUrl;
	private long bpSize = 0;
	private List<TF2Item> items;
	private static final Double REFINED_PRICE = 0.35;

	public Backpack(String id) {
		apiUrl = String.format("http://api.steampowered.com/IEconItems_440/GetPlayerItems/v0001/?key=%s&format=json&SteamID=%s",
				Configuration.API_KEY, id);
		items = new LinkedList<TF2Item>();
	}

	public boolean init() throws ParseException {
		JSONParser parser = new JSONParser();
		String jsonResponse = super.getJson(apiUrl);
		if(jsonResponse == null) {
			return false;
		}
		Object responseObj = parser.parse(jsonResponse);
		JSONObject result = (JSONObject)((JSONObject) responseObj).get("result");

		bpSize = (long)result.get("num_backpack_slots");

		JSONArray itemsArray = (JSONArray) result.get("items");

		for(int i = 0 ; i < itemsArray.size() ; i++) {
			JSONObject it = (JSONObject) itemsArray.get(i);
			ItemQuality cQuality = ItemQuality.NORMAL;
			switch((int)(long)it.get("quality")) {
				case 0: cQuality = ItemQuality.NORMAL; break;
				case 1: cQuality = ItemQuality.GENUINE; break;
				case 3: cQuality = ItemQuality.VINTAGE; break;
				case 5: cQuality = ItemQuality.UNUSUAL; break;
				case 6: cQuality = ItemQuality.UNIQUE; break;
				case 8: cQuality = ItemQuality.VALVE; break;
				case 11: cQuality = ItemQuality.STRANGE; break;
				case 13: cQuality = ItemQuality.HAUNTED; break;
			}
			items.add(new TF2Item("N/A", (long)it.get("defindex"), cQuality));
		}

		return true;
	}

	public boolean hasItem(long itemId, ItemQuality quality) {
		return items.contains(new TF2Item("N/A", itemId, quality));
	}

	public boolean hasUnusual() {
		for(TF2Item item : items) {
			if(item.getQuality().equals(ItemQuality.UNUSUAL)) {
				return true;
			}
		}
		return false;
	}

	public long getSize() {
		return bpSize;
	}

	public static Double getValue(String id) throws ParseException {
		String api = String.format("http://backpack.tf/api/IGetUsers/v2/?steamids=%s&format=json", id);
		String jsonResponse = Backpack.getJson(api);
		if(jsonResponse == null) {
			return 0.0;
		}
		JSONParser parser = new JSONParser();
		Object responseObj = parser.parse(jsonResponse);
		JSONObject response = (JSONObject)((JSONObject) responseObj).get("response");
		JSONObject players = (JSONObject)response.get("players");
		JSONObject player = (JSONObject)players.get("0");

		if(player.get("backpack_value").getClass().equals(Long.class)) {
			return (((double)(long)player.get("backpack_value")) * REFINED_PRICE);
		} else {
			return (((double)player.get("backpack_value")) * REFINED_PRICE);
		}
	}
}

File: tf2-item-scanner/tf2-item-scanner/engine/Backpack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using tf2_item_scanner.utils;
using Newtonsoft.Json.Linq;
using System.Net;
using System.IO;

namespace tf2_item_scanner.engine
{
    class Backpack
    {
        private string _id;
        private long _size;
        private string _apiUrl;
        private Utils utils;

        private List<TF2Item> _items;

        public Backpack(string id, Utils utils)
        {
            _apiUrl = utils.BackpackUrl + id;
            _items = new List<TF2Item>();
            _id = id;
            this.utils = utils;
        }

        public bool Init()
        {
            try
            {
                string json = utils.GetJson(_apiUrl);

                JObject o = JObject.Parse(json);
                JObject result = (JObject)o["result"];

                _size = (long)result["num_backpack_slots"];

                JArray itemsArray = (JArray)result["items"];

                for (int i = 0; i < itemsArray.Count; i++)
                {
                    JObject it = (JObject)itemsArray[i];
                    ItemQuality quality = ItemQuality.Normal;

                    switch ((int)it["quality"])
                    {
                        case 0: quality = ItemQuality.Normal; break;
                        case 1: quality = ItemQuality.Genuine; break;
                        case 3: quality = ItemQuality.Vintage; break;
                        case 5: quality = ItemQuality.Unusual; break;
                        case 6: quality = ItemQuality.Unique; break;
                        case 8: quality = ItemQuality.Valve; break;
                        case 11: quality = ItemQuality.Strange; break;
                        case 13: quality = ItemQuality.Haunted; break;
                    }
                    _items.Add(new TF2Item((string)it["name"], (long)it["defindex"], quality));
                }

                return true;
            }
            catch (Exception e)
            {
                return false;
            }
        }

        public bool HasItem(TF2Item item)
        {
            return _items.Where(x => x.DefinitionIndex == item.DefinitionIndex &&
                x.Quality == item.Quality).Count() > 0;
        }

        public TF2Item HasUnusual()
        {
            IEnumerable<TF2Item> found =  _items.Where(x => x.Quality == ItemQuality.Unusual 
                && x.DefinitionIndex != 267 && x.DefinitionIndex != 266);
            return found.Count() == 0 ? null : found.First();
        }

        #region PROPERTIES

        public long Size
        {
            get { return _size; }
        }

        public long ItemsCount
        {
            get { return _items.Count; }
        }

        public double Value
        {
            get
            {
                string api = utils.ValueUrl + _id;
                string json = utils.GetJson(api);

                try
                {
                    JObject o = JObject.Parse(json);
                    JObject result = (JObject)o["response"];
                    JObject players = (JObject)result["players"];
                    JObject player = (JObject)players["0"];

                    return ((double)player["backpack_value"]) * utils.RefPrice;
                }
                catch (Exception e)
                {
                    return 0.0;
                }
            }
        }

        #endregion
    }
}
